<a href="https://colab.research.google.com/github/adong-hood/cs200/blob/main/ch5_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd


## Exploring the happiness data

The World Happiness Report is a landmark survey of the state of global happiness.

Read in data as Pandas data frame

In [ ]:
happiness_df = pd.read_csv('http://pluto.hood.edu/~dong/datasets/happiness_2017.csv')
happiness_df.head(2)

Country          Region  Rank  HappinessScore  Life Ladder  \
0   Norway  Western Europe     1           7.537     7.578745   
1  Denmark  Western Europe     2           7.522     7.593702   

   Log GDP per capita  Social support  Healthy life expectancy at birth  \
0           11.081789        0.950128                         71.086586   
1           10.748989        0.952100                         71.662498   

   Freedom to make life choices  Generosity  Perceptions of corruption  \
0                      0.953017    0.210104                   0.249711   
1                      0.955416    0.145387                   0.181148   

   Positive affect  Negative affect  Confidence in national government  
0         0.849100         0.202914                           0.717160  
1         0.823667         0.205775                           0.572353

In [ ]:
happiness_df.shape

(140, 14)

Fourteen columns of data for over 140 countries.

In [ ]:
happiness_df.columns

Index(['Country', 'Region', 'Rank', 'HappinessScore', 'Life Ladder',
       'Log GDP per capita', 'Social support',
       'Healthy life expectancy at birth', 'Freedom to make life choices',
       'Generosity', 'Perceptions of corruption', 'Positive affect',
       'Negative affect', 'Confidence in national government'],
      dtype='object')

In [ ]:
happiness_df.isnull().sum()

Country                               0
Region                                0
Rank                                  0
HappinessScore                        0
Life Ladder                           0
Log GDP per capita                    7
Social support                        1
Healthy life expectancy at birth      0
Freedom to make life choices          1
Generosity                            8
Perceptions of corruption            12
Positive affect                       1
Negative affect                       1
Confidence in national government    13
dtype: int64

In [ ]:
happiness_df = pd.read_csv('http://pluto.hood.edu/~dong/datasets/happiness_2017.csv').dropna(axis=1, how='any')
happiness_df.head(2)

Country          Region  Rank  HappinessScore  Life Ladder  \
0   Norway  Western Europe     1           7.537     7.578745   
1  Denmark  Western Europe     2           7.522     7.593702   

   Healthy life expectancy at birth  
0                         71.086586  
1                         71.662498

<p>By default, dropna() will drop all rows in which any null value is present. Alternatively, you can drop NA values along a different axis; axis=1 drops all columns containing a null value.</p>
<p>The default is how='any', such that any row or column (depending on the axis keyword) containing a null value will be dropped. You can also specify how='all', which will only drop rows/columns that are all null values.</p>

In [ ]:
happiness_df.shape

(140, 6)

only 6 columns left.

In [ ]:
happiness_df.columns

In [ ]:
happiness_df.describe()

Rank  HappinessScore  Life Ladder  Healthy life expectancy at birth
count  140.0000      140.000000   140.000000                        140.000000
mean    70.5000        5.398907     5.491702                         63.440981
std     40.5586        1.109750     1.122178                          7.595597
min      1.0000        2.693000     2.661718                         44.386528
25%     35.7500        4.542500     4.636694                         58.005130
50%     70.5000        5.302000     5.586507                         65.209110
75%    105.2500        6.212000     6.278276                         69.062944
max    140.0000        7.537000     7.788252                         76.536362

In [ ]:
happiness_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140 entries, 0 to 139
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Country                           140 non-null    object 
 1   Region                            140 non-null    object 
 2   Rank                              140 non-null    int64  
 3   HappinessScore                    140 non-null    float64
 4   Life Ladder                       140 non-null    float64
 5   Healthy life expectancy at birth  140 non-null    float64
dtypes: float64(3), int64(1), object(2)
memory usage: 6.7+ KB


#### 4.1 What country has the highest happiness score?

In [ ]:
happiness_df[happiness_df['HappinessScore'] == happiness_df['HappinessScore'].max()]['Country'].iloc[0]

'Norway'

In [ ]:
happiness_df.columns

Index(['Country', 'Region', 'Rank', 'HappinessScore', 'Life Ladder',
       'Log GDP per capita', 'Social support',
       'Healthy life expectancy at birth', 'Freedom to make life choices',
       'Generosity', 'Perceptions of corruption', 'Positive affect',
       'Negative affect', 'Confidence in national government'],
      dtype='object')

In [ ]:
happiness_df['Region'].unique()

array(['Western Europe', 'North America and ANZ',
       'Middle East and North Africa', 'Latin America and Caribbean',
       'Central and Eastern Europe', 'Southeast Asia', 'East Asia',
       'Commonwealth of Independent States', 'Sub-Saharan Africa',
       'South Asia'], dtype=object)

#### 4.2 What countries from North America and ANZ have happiness score higher than the average?

In [ ]:
happiness_df[(happiness_df['Region'] == 'North America and ANZ') & (happiness_df['HappinessScore'] > happiness_df['HappinessScore'].mean())]['Country']

6       New Zealand
7         Australia
12    United States
Name: Country, dtype: object

#### 4.3 Find the first 10 countries with the highest life ladder.

In [ ]:
happiness_df.sort_values(by='Life Ladder', ascending=False)[['Country', 'Life Ladder']].head(10)

Country  Life Ladder
4       Finland     7.788252
1       Denmark     7.593702
0        Norway     7.578745
2       Iceland     7.476214
3   Switzerland     7.473593
5   Netherlands     7.458965
9        Israel     7.331036
6   New Zealand     7.327183
11      Austria     7.293728
8        Sweden     7.286805

### 4.3 Look up happiness scores by country.

### 4.4 Find the number of country names starting with B or C.
Slicing based on string comparison.

In [ ]:
happiness_df.index

RangeIndex(start=0, stop=140, step=1)

In [ ]:
lookup_by_country = happiness_df.set_index('Country').sort_index()
lookup_by_country.head(2)

Region  Rank  HappinessScore  Life Ladder  \
Country                                                                      
Afghanistan                  South Asia   129           3.794     2.661718   
Albania      Central and Eastern Europe   100           4.644     4.639548   

             Log GDP per capita  Social support  \
Country                                           
Afghanistan            7.460144        0.490880   
Albania                9.373718        0.637698   

             Healthy life expectancy at birth  Freedom to make life choices  \
Country                                                                       
Afghanistan                         52.339527                      0.427011   
Albania                             69.051659                      0.749611   

             Generosity  Perceptions of corruption  Positive affect  \
Country                                                               
Afghanistan    -0.10634                   0.954393         0.496349   
Albania        -0.03514                   0.876135         0.669241   

             Negative affect  Confidence in national government  
Country                                                          
Afghanistan         0.371326                           0.261179  
Albania             0.333884                           0.457738

In [ ]:
lookup_by_country.index

Index(['Afghanistan', 'Albania', 'Algeria', 'Argentina', 'Armenia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh',
       ...
       'Ukraine', 'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Uzbekistan', 'Vietnam', 'Yemen', 'Zambia', 'Zimbabwe'],
      dtype='object', name='Country', length=140)

In [ ]:
begin = lookup_by_country[lookup_by_country.index.str.startswith('B')].index[0]

In [ ]:
end = lookup_by_country[lookup_by_country.index.str.startswith('C')].index[-1]

In [ ]:
lookup_by_country.loc[begin:end].shape

(24, 13)

We can use the `startswith` predicate (in the `str` attribute) to find all countries that start with a particular string or letter.

## Your work for Chapter 5, Section 5.4 starts here.

There are three questions in 5.4. Include another question of your own.